src/geometry/rtree.cpp

```diff
@@ -252,32 +252,41 @@ std::size_t RTree<cs>::report() const noexcept {
 
 template <typename cs>
 template <typename Predicate, typename Filter>
-auto RTree<cs>::apply_predicate(Predicate&& predicate, Filter&& filter) const -> query_results {
-  return rtree_.qbegin(std::forward<Predicate>(predicate) &&
+auto RTree<cs>::apply_intersect_predicate(Predicate&& predicate, Filter&& filter) const -> query_results {
+  return rtree_.qbegin(predicate &&
                        bgi::satisfies([this, f = std::move(filter)](doc_type const& doc) -> bool {
                          return lookup(doc).map(f).value_or(false);
                        }));
 }
+template <typename cs>
+template <typename Predicate, typename Filter>
+auto RTree<cs>::apply_union_predicate(Predicate&& predicate, Filter&& filter) const -> query_results {
+  return rtree_.qbegin(bgi::satisfies([this, p = std::move(predicate), f = std::move(filter)](doc_type const& doc) -> bool {
```

Comment:
Maybe:
p = std::forward(predicate)
and 
f = std::forward(filter)
?
Since predicate and filter can be universal references, depending on the template initialization.

Reply:
Here I specifically want these to be moves. in fact, these functions should not accept by universal reference at all. I want the lambda to take ownership of the predicates, leaving no option for potentially dangling refs. Which it won't do if forwarding an lvalue ref. Not that that occurs here though, as all calls to these functions are made with temporaries. Nevertheless, for future proofing, we don't want to pass a reference to the predicate, we specifically want to copy the predicate.

@oshadmi this applies to the above as well. But `rtree.qbegin` accepts its predicate by `const&` so we gain nothing by doing so. I knew there was a reason why it didn't matter there :sweat_smile: